src/data/ordmap/ordnode.lean

```diff
@@ -0,0 +1,1152 @@
+/-
+Copyright (c) 2017 Mario Carneiro. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro
+-/
+import data.list.defs
+import data.nat.psub
+
+/-!
+# Ordered sets
+
+This file defines a data structure for ordered sets, supporting a
+variety of useful operations including insertion and deletion,
+logarithmic time lookup, set operations, folds,
+and conversion from lists.
+
+The `ordnode α` operations all assume that `α` has the structure of
+a total preorder, meaning a `≤` operation that is
+
+* Transitive: `x ≤ y → y ≤ z → x ≤ z`
+* Reflexive: `x ≤ x`
+* Total: `x ≤ y ∨ y ≤ x`
+
+For example, in order to use this data structure as a map type, one
+can store pairs `(k, v)` where `(k, v) ≤ (k', v')` is defined to mean
+`k ≤ k'` (assuming that the key values are linearly ordered).
+
+Two values `x,y` are equivalent if `x ≤ y` and `y ≤ x`. An `ordnode α`
+maintains the invariant that it never stores two equivalent nodes;
+the insertion operation comes with two variants depending on whether
+you want to keep the old value or the new value in case you insert a value
+that is equivalent to one in the set.
+
+The operations in this file are not verified, in the sense that they provide
+"raw operations" that work for programming purposes but the invariants
+are not explicitly in the structure. See `ordset` for a verified version
+of this data structure.
+
+## Main definitions
+
+* `ordnode α`: A set of values of type `α`
+
+## Implementation notes
+
+Based on weight balanced trees:
+
+ * Stephen Adams, "Efficient sets: a balancing act",
+   Journal of Functional Programming 3(4):553-562, October 1993,
+   <http://www.swiss.ai.mit.edu/~adams/BB/>.
+ * J. Nievergelt and E.M. Reingold,
+   "Binary search trees of bounded balance",
+   SIAM journal of computing 2(1), March 1973.
+
+Ported from Haskell's `Data.Set`.
+
+## Tags
+
+ordered map, ordered set, data structure
+
+-/
+
+universes u
+
+/-- An `ordnode α` is a finite set of values, represented as a tree.
+  The operations on this type maintain that the tree is balanced
+  and correctly stores subtree sizes at each level. -/
+inductive ordnode (α : Type u) : Type u
+| nil {} : ordnode
+| node (size : ℕ) (l : ordnode) (x : α) (r : ordnode) : ordnode
+
+namespace ordnode
+variable {α : Type u}
+
+instance : has_emptyc (ordnode α) := ⟨nil⟩
+instance : inhabited (ordnode α) := ⟨nil⟩
+
+/-- (Internal use only.)
```

Comment:
Could you standardize the location and formatting of the "internal use only" markers in these doc strings? I'd go for something like
```suggestion
/-- **Internal use only**

```
Note the extra line break.

Does it break things to make these internal definitions private?

Reply:
> Does it break things to make these internal definitions private?

I've discussed why I think it is best to avoid `private` definitions under all circumstances elsewhere. What I really mean to say here is that these are functions that don't maintain the module invariants; they would be `unsafe` functions in rust-speak but I don't know that we have a good terminology for them. They are just fine to be used by downstream users but doing so probably means you are working "under" the ordmap API. A casual user doesn't need to ever know about these functions so it would be nice to hide them in documentation or set them off somehow, hence the use of `Internal use only.` in the doc strings. A more standardized approach to this would be great.